### ISSUE TITLE
Add context management for workflow submission

### ISSUE THREAD

**Issue by flaviuvadan** [argoproj-labs/hera-workflows#79]
Currently, users have to add tasks to a workflow independently. After a task is defined, it is necessary to call `workflow.add_tasks(t1, t2, ...)` for otherwise the tasks are not added to the template of the DAG in the main workflow. To save some effort in this arena, a context manager can be introduced to auto-insert tasks into a workflow, and submit a workflow during the exit phase of the context. This will provide two ways of adding tasks to a workflow. 

Current, and supported way of adding tasks and submitting a workflow:
```python
wf = Workflow('test')
t1 = Task('t1', image='docker/whalesay', command=['cowsay', 'foo'])
t2 = Task('t2', image='docker/whalesay', command=['cowsay', 'foo'])
t1 >> t2
wf.add(t1, t2)
wf.submit()
```

Potential future (second) way of adding tasks and submitting a workflow:
```python
with Workflow('test') as wf:
    Task('t1', image='docker/whalesay', command=['cowsay', 'foo']) 
    >> Task('t2', image='docker/whalesay', command=['cowsay', 'foo'])
```

Alternatively:
```python
with Workflow('test') as wf:
    t1 = Task('t1', image='docker/whalesay', command=['cowsay', 'foo']) 
    t2 = Task('t2', image='docker/whalesay', command=['cowsay', 'foo'])
    t1 >> t2
```

This will require the implementation of a context manager on `Workflow` and a container for `Task`, which tasks insert themselves into during `__init__`:

```python
class Workflow:

    def __enter__(self, ...):
        ...
        return self

    def __exit__(self, ...):
        ...
        return self.submit()

class _Container(list):
    
    @classmethod
    def add(cls, t: Task):
        cls.append(t)

class Task:

    def __init__(...):
        ...
        _Container.add(self)
```

Questions to consider:
* Does this solve a real problem or would it cause confusion because of of the implicit behavior of task additions?
* Is this an interface that abides by Hera's principle of simplicity? Is the `_Container` usage a bit too abstract?
* Is it safe to assume that exiting a context should indeed submit a workflow? If not but the design is something desirable, should users be provided with the feature, along with a parameter on `Workflow` that says "auto_submit`? Does that pollute the interface of `Workflow` unnecessarily with QOL features rather than focusing on exposing Argo specific parameters?

Would love to discuss this @bchalk101 and @bobh66! Any thoughts? Would this be useful to your use case?

**Comment by dixon1e** [comment]
@flaviuvadan Very interested to hear thoughts on this. While the current way is clear, the scientists on one of our teams were put off by having to add the jobs directly. I cannot answer why, it is not obvious to me where the resistance is. 

Having said that, to my own eye, the "alternate" method looks nicer. Your concerns are valid, that it must be clear and obvious (in some way) that an auto-submit will happen upon context exit.

**Comment by flaviuvadan** [comment]
> scientists on one of our teams were put off by having to add the jobs directly. I cannot answer why, it is not obvious to me where the resistance is.

This is interesting to hear, and I can relate! It's the primary motivation behind Hera, really... to make it incredibly easy to focus on the stuff a scientist cares about, which is not task/workflow definition/construction but the scientific code. 

I think we certainly need to strike a balance between explicit and implicit. Perhaps the auto-insert of tasks into a workflow is the "right" level of abstraction in this case? The implicit nature of task auto-insertion may "just make sense" for people using the context manager (i.e "I write a task inside a workflow context and I expect it to be added"). By comparison, auto-submission may not be as welcome/may act as a surprise (i.e "Oh, I was not done defining my tasks and this thing submitted?! What do I do now?!"). In this case, perhaps auto-adding using a container inside a `Task.__init__` is what we want and expect people to run `workflow.submit()` within the context?

**Comment by dixon1e** [comment]
Some humans love to push a BIG RED BUTTON to make a machine go, and the `submit` may represents the action they want. On the `add tasks` observation, my belief is this is more about chaining workflows. It may not be clear to someone who is used to workflow templates, exactly what the comparison is to tasks.

**Comment by flaviuvadan** [comment]
> the submit may represent the action they want

Yea, I think we're aligned here. The `submit`, as you said, represents the clear action here (an action users _care_ about), whereas adding tasks to a workflow after defining them is not something they may _care_ about.

**Comment by bobh66** [comment]
I like this idea, it seems like a cleaner flow, I could use it for one of the specific scenarios I have with CronWorkflow.

This does make me think about the default task handling logic - if a user adds a list of Tasks to the workflow, with no ordering logic, they will (I believe) execute in parallel.   If a (non-expert) user might think that if they go to the trouble of creating a list of Tasks in a specific order that the Workflow would honor that execution order "automatically".  I'm not saying to change anything, I'm just wondering if this is a valid user expectation that might be lurking?  If someone has a "large" number of Tasks, maybe programmatically defined, that they hand off to a Workflow, does it make sense to have an option to execute them serially instead of in parallel?

Definitely not something we need now, but maybe in the future?

Otherwise LGTM 👍

**Comment by flaviuvadan** [comment]
> This does make me think about the default task handling logic

@bobh66 this is a good point. Correct me if I am wrong, but you're saying it's possible that the definition order has implications on the dependency chain of tasks. So, something like:

```python
with Workflow() as w:
  t1 = Task(...)
  t2 = Task(...)
```
is equivalent to and implies:
```python
w = Workflow()
t1 = Task(...)
t2 = Task(...)
t1 >> t2
w.add_tasks(t1, t2)
w.submit()
```

If so, this is something that's very interesting. I wonder, though, if some users might want to run something like a daemon task alongside other tasks that are chained, but have to communicate with the daemon task, or something like that. I can envision use cases where there's this general pattern of launching a task parallel to a chain and the singular task acts as a "watcher" of some kind, or a "server" (depending on the communication direction). 

I think I am very open to this idea, if it is indeed what you are saying, but I'd prefer we separate that development from this one. Technically it's simple but the implications are quite serious in light of the minor technical involvement. I'll open a new issue with this idea after we develop this one, should we go ahead with it, and perhaps have a separate discussion there?

**Comment by bchalk101** [comment]
The context manager looks like a good idea to me, specifically from the perspective of adding the tasks automatically to a workflow. However, I believe that a `submit()` method is important.

> represents the clear action here (an action users care about)

From the perspective of the clear action, but also from being able to use `Workflows` as templated code, I may want to construct a workflow and then reuse it or share it, without it being submitted.

> Is the _Container usage a bit too abstract?

I can see an issue here, currently, the internals of Hera are simple and there is a small learning curve to be able to contribute, the introduction of a certain level of implicit/convoluted flow, can trip people up and potentially negate the benefit it would give the Hera end user.

**Comment by flaviuvadan** [comment]
> may want to construct a workflow and then reuse it or share it, without it being submitted.

Thanks for sharing this great example of a use-case that supports making `submit()` an explicit action 🙂 

> can trip people up and potentially negate the benefit it would give the Hera end user.

I can definitely see this. Perhaps extensive documentation is the way to approach this? Docstrings and diagrams. Also, I should clarify that the introduction of the context manager does not imply the removal of `add_task/s`! I suggest Hera offers both approaches as both have their own advantages! Providing `add_task/s` hopefully empowers users to stitch together tasks in very convoluted ways whereas the context manager is more oriented towards giving easy access to constructing simpler workflows 🙂

@bchalk101 any thoughts on the auto-dependency chaining aspect illustrated [here](https://github.com/argoproj-labs/hera-workflows/issues/79#issuecomment-1030319545)?

**Comment by bchalk101** [comment]
While it is very succinct, I feel it could lead to a lot of confusion and unexpected behavior. 

>   if some users might want to run something like a daemon task alongside other tasks that are chained

We use daemoned steps often, specifically for training where we have a data service serving the training workers, and it would remove for us all the benefit of the context management.

**Comment by bobh66** [comment]
I think the magic _Container is confusing, or at least it's not obvious to the user how creating the Task inside the context manager automagically adds the Task to the Workflow.

I think something like this would be more readable and more intuitive to the user:

```
with Workflow('test') as wf:
    wf.add_task(Task('t1', image='docker/whalesay', command=['cowsay', 'foo']))
```

This makes it more explicit how the Task is being associated with the Workflow.

I don't mind the automatic submit() happening, because that's kind of the point to using the context manager.  If I didn't want the automatic submit() to happen I could write it the regular way.

I think the automatic chaining (or lack thereof) only becomes an issue if you are doing a loop inside the context manager and you are adding a bunch of tasks that you want chained in a certain way.   You can certainly leave the chaining out and have them all run in parallel, or you can figure some programmatic way to chain them the way you want.

If you are just defining a fixed set of tasks within the context manager then you can also define the chaining to happen (or not) the way you want them to.

Maybe at some point in the future we might want to add a para,meter to add_tasks() that could control the chaining applied to a list of tasks, but I don't think that's needed now.

**Comment by bobh66** [comment]
@flaviuvadan I have come across a scenario that raises the issue of the default task handling being parallel instead of sequential.

I have a scenario where I would like to define a generic Workflow object that always adds a "start" indication task as part of the __init__ method,, which sends an HTTP status message to a server, and always adds a "complete" indication task which gets added by the submit() method just before calling super().submit().   The tasks "between" the start/end tasks can be organized however the user wants, but I would like the start task to always run first and the completed task to always run last.

I'm not exactly sure how to implement this, since it really means making the default task creation flow sequential instead of parallel (or requires an option to signal that it should be sequential).   It also separates the Task creation/handling, and the "payload" tasks would not "know" that the start/complete tasks exist, so it may not even be possible to do what I'm proposing.

I may be better off looking at Argo Events to do what I'm looking for, but I wanted to start with Hera before diving into Argo Events.

Any thoughts or ideas/suggestions are greatly appreciated.

Thanks!

**Comment by flaviuvadan** [comment]
@bobh66 

> I think the magic _Container is confusing, or at least it's not obvious to the user how creating the Task inside the context manager automagically adds the Task to the Workflow.

This is interesting. I view the `Container` as more relevant to developers/contributors rather than the end users 🤔 A lot of frameworks follow this pattern of having some abstract collection in the background that gathers any instantiated objects, which are then used for submission. For instance, Airflow and Prefect follow a similar pattern - declare inside a context (tasks inside the `with`), execute outside of it (`.submit` on the workflow object). 

> This makes it more explicit how the Task is being associated with the Workflow

Yea, this is the reasoning that was used when `add_task`/`add_tasks` was implemented! The freedom to add tasks in whatever order is quite powerful. To be clear, though, the introduction of the context manager would not remove these, so users have the freedom to either use the context manager if the tasks are perfectly sequential, or use the current approach for declaring, adding, and submitting tasks! 🙂 

> I have a scenario where I would like to define a generic Workflow object that always adds a "start" indication task as part of the init method,, which sends an HTTP status message to a server, and always adds a "complete" indication task which gets added by the submit() method just before calling super().submit(). The tasks "between" the start/end tasks can be organized however the user wants, but I would like the start task to always run first and the completed task to always run last.

Oh, this is a cool scenario! What do you think about overwriting the `submit` method of the workflow in the code you are working with, and perhaps use the `add_head` and `add_tail` at submission time? That way Hera can add a task at the beginning of the workflow that will perform your HTTP call, and add a task at the end of the workflow that can run last! This is a solution one could use to have Hera back the workflow exclusively. If this is not sufficient, perhaps we can consider introducing lifecycle hooks for workflow events, such as "running" and "failed", along with template/task specifications to run. There is also an exit handler on workflows that can be used to invoke a specific template upon exit. If the suggestion I wrote above does not solve your problem, then please submit some issues with the requested features! Another pretty exciting idea is to bring Events into Hera 🙂 that's a big discussion we could have but we should have it on a different issue!

**Comment by bobh66** [comment]
>  I view the Container as more relevant to developers/contributors rather than the end user

I agree, I just view it as "magic" that is doing something potentially unexpected by the user, or at least not obvious to the user that it's happening.

One question - since _Container is global and uses a classmethod add(), are there potentially problems with mixing up collections of Tasks when running async?  I guess the question is really whether there is a separate instance of _Container for each Workflow, to keep the tasks separated so they don't get tasks from multiple workflows mixed together? 

>  What do you think about overwriting the submit method of the workflow in the code you are working with, and perhaps use the add_head and add_tail at submission time?

Thanks - that's exactly what I needed.  I had missed the add_head/tail methods.

**Comment by flaviuvadan** [comment]
> One question - since _Container is global and uses a classmethod add(), are there potentially problems with mixing up collections of Tasks when running async? I guess the question is really whether there is a separate instance of _Container for each Workflow, to keep the tasks separated so they don't get tasks from multiple workflows mixed together?

This is a very good point, actually, and affect the design of the solution. Under what circumstances could this happen? I am trying to think about APIs that launch workflows, or something similar, but those are probably thread safe? Each thread will have its own running container that accumulates tasks and is emptied upon submission.

> that's exactly what I needed

🎉

**Comment by flaviuvadan** [comment]
> If there is a _Container instance dedicated to the Workflow object instance then this would not be a problem.

These are great examples, thank you for sharing 🙂 what do you think could be a way to design this without imposing knowledge about a workflow onto a task?

**Comment by Trollgeir** [comment]
> For instance, Airflow and Prefect follow a similar pattern - declare inside a context (tasks inside the with), execute outside of it (.submit on the workflow object).

I find this to be one of the most compelling arguments (given that there aren't any clear downsides). There's already seems to be a standard for this, meaning it's an easier transition for people with knowledge in these tools to adapt to Hera.

**Comment by flaviuvadan** [comment]
> I find this to be one of the most compelling arguments (given that there aren't any clear downsides). There's already seems to be a standard for this, meaning it's an easier transition for people with knowledge in these tools to adapt to Hera.

Thank you for the support, will prioritize for 3.6.0 :)

**Comment by flaviuvadan** [comment]
https://github.com/argoproj-labs/hera-workflows/pull/209
